### DISCUSSION TITLE
How to get IP address to pass to htonl()

### DISCUSSION THREAD

**Discussion by krishnak** [micropython/micropython#9742] [Core Development] [answered]
I am basically copying the code inside esp_ifconfig  in modnetwork.c to a new method to enable NAT 

```
wlan_if_obj_t *self = MP_OBJ_TO_PTR(args[0]);
    tcpip_adapter_ip_info_t info;
    tcpip_adapter_dns_info_t dns_info;
    tcpip_adapter_get_ip_info(self->if_id, &info);
    tcpip_adapter_get_dns_info(self->if_id, DNS_MAIN, &dns_info);

```    
the ifconfig is set to to 192.168.5.1 on the AP_IF - how do I get this value from the info variable to pass it to the below function?

ip_napt_enable(htonl(napt_netif_ip), 1);

I have tried to print the IP address from the variable info, but it is not the right value
```

    mp_printf(&mp_plat_print, "IP address is: %lu\n",info.ip.addr);   
    mp_printf(&mp_plat_print, "IP address is: %lu\n",info.ip);   

```
It prints out a value of 395811, where as the correct value should be 3232236801

**Comment by karfas** [comment]
What has this to do with micropython ?
All these are ESP-IDF or lwIP calls; you might want to consult the ESP-IDF documentation for details of the data returned from e.g. `tcpip_adapter_get_ip_info`.

**Reply by krishnak** [reply]
The reason for me asking here is because there appears to be some relation to the method signature declared here in the object getting the right instance

MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(esp_nat_enable_obj,1,2, esp_nat_enable);

if I use
MP_DEFINE_CONST_FUN_OBJ_1(esp_nat_enable_obj, esp_nat_enable); - it doesn't work

**Comment by krishnak** [comment, marked as answer]
Just for anyone who may be interested in enabling NAT on their ESP32

Add this to your modenetwork.c located in /ports/esp32
```

#include "lwip/lwip_napt.h"

STATIC mp_obj_t esp_nat_enable(size_t n_args,const mp_obj_t *args){
    wlan_if_obj_t *self = MP_OBJ_TO_PTR(args[0]);
        if(self->if_id!=1)
        return;
    tcpip_adapter_ip_info_t info;
    tcpip_adapter_dns_info_t dns_info;
    tcpip_adapter_get_ip_info(self->if_id, &info);
    
    tcpip_adapter_get_dns_info(self->if_id, DNS_MAIN, &dns_info);
    
    u32_t *dns_ip = (u32_t *)&dns_info.ip;;
    
    u32_t *ip = (u32_t *)&info.ip;
    ip_addr_t dnsserver;

    dhcps_offer_t dhcps_dns_value = OFFER_DNS;
    dhcps_set_option_info(6, &dhcps_dns_value, sizeof(dhcps_dns_value));
    
    dnsserver.u_addr.ip4.addr = *dns_ip;
    dnsserver.type = IPADDR_TYPE_V4;
    dhcps_dns_setserver(&dnsserver);

    ip_napt_enable(*ip, 1);
    return mp_const_none;
}
MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(esp_nat_enable_obj,1,2, esp_nat_enable);

```

Add this line to modnetwork.h

`MP_DECLARE_CONST_FUN_OBJ_VAR_BETWEEN(esp_nat_enable_obj);
`
Make sure your network_wlan.c looks like this
```
STATIC const mp_rom_map_elem_t wlan_if_locals_dict_table[] = {
    { MP_ROM_QSTR(MP_QSTR_active), MP_ROM_PTR(&network_wlan_active_obj) },
    { MP_ROM_QSTR(MP_QSTR_connect), MP_ROM_PTR(&network_wlan_connect_obj) },
    { MP_ROM_QSTR(MP_QSTR_disconnect), MP_ROM_PTR(&network_wlan_disconnect_obj) },
    { MP_ROM_QSTR(MP_QSTR_status), MP_ROM_PTR(&network_wlan_status_obj) },
    { MP_ROM_QSTR(MP_QSTR_scan), MP_ROM_PTR(&network_wlan_scan_obj) },
    { MP_ROM_QSTR(MP_QSTR_isconnected), MP_ROM_PTR(&network_wlan_isconnected_obj) },
    { MP_ROM_QSTR(MP_QSTR_config), MP_ROM_PTR(&network_wlan_config_obj) },
    { MP_ROM_QSTR(MP_QSTR_ifconfig), MP_ROM_PTR(&esp_ifconfig_obj) },
    { MP_ROM_QSTR(MP_QSTR_nat_enable), MP_ROM_PTR(&esp_nat_enable_obj) },
};
```
```

CONFIG_LWIP_IP_FORWARD=y
CONFIG_LWIP_IPV4_NAPT=y

these config parameters need to be present in sdkconfig.base under /esp32/boards
```

Once you build it and flash it, connect STA to a working router, set your ifconfig parameters on AP_IF correctly to include a DNS server for ex ('192.168.5.1','255.255.255.0','192.168.5.1','8.8.8.8'). Then you can call network.WLAN(network.AP_IF).nat_enable() and any devices connected to ESP32 AP will be able connect to the internet.

**Comment by stephanelsmith** [comment]
This would be an really awesome feature to include, pull request?

**Reply by krishnak** [reply]
My code base is a clone of glenn who has implemented ESPNow whose pull request itself is pending. My version control experience with git is pretty limited. It may be easier for the maintainers to copy the stuff and build and test it, as it is a very small change involving 4 files.

**Reply by stephanelsmith** [reply]
I would **LOVE** t see Glenn's ESPNow implementation brought in officially.  I'm hesitant to bring that into my project especially if it's a side-repo.  I had some difficulty getting simple S3 related pin updates merged in, so I'm in the same boat regarding git.  Regardless, thank you very much for documenting the changes you made here.  This is very cool!
